protocol-tokio/src/protocol/codec/message.rs

```diff
@@ -88,17 +89,113 @@ pub enum Message {
+
+impl<B: server::block::BlockService, T: server::transaction::TransactionService> AsNtEvent
+    for ChainMessage<B, T>
+where
+    <B as server::block::BlockService>::BlockId: cbor_event::Deserialize,
+    <B as server::block::BlockService>::BlockId: cbor_event::Serialize,
+    <B as server::block::BlockService>::Block: cbor_event::Deserialize,
+    <B as server::block::BlockService>::Block: cbor_event::Serialize,
+    <B as server::block::BlockService>::Header: cbor_event::Deserialize,
+    <B as server::block::BlockService>::Header: cbor_event::Serialize,
+    <T as server::transaction::TransactionService>::TransactionId: cbor_event::Serialize,
+    <T as server::transaction::TransactionService>::TransactionId: cbor_event::Deserialize,
+{
+    fn to_nt(self) -> nt::Event {
+        use self::nt::Event::*;
+        match self {
+            ChainMessage::GetBlockHeaders(lwcid, gbh) => {
+                Data(lwcid, MessageType::MsgGetHeaders.encode_with(&gbh))
+            }
+            ChainMessage::BlockHeaders(lwcid, bh) => Data(lwcid, cbor!(&bh).unwrap().into()),
+            ChainMessage::GetBlocks(lwcid, gb) => {
+                Data(lwcid, MessageType::MsgGetBlocks.encode_with(&gb))
+            }
+            ChainMessage::Block(lwcid, b) => Data(lwcid, cbor!(&b).unwrap().into()),
+            ChainMessage::SendTransaction(lwcid, tx) => Data(lwcid, cbor!(&tx).unwrap().into()),
+            ChainMessage::TransactionReceived(lwcid, rep) => {
+                Data(lwcid, cbor!(&rep).unwrap().into())
+            }
+            ChainMessage::Subscribe(lwcid, keep_alive) => {
+                let keep_alive: u64 = if keep_alive { 43 } else { 42 };
+                Data(lwcid, MessageType::MsgSubscribe1.encode_with(&keep_alive))
+            }
+        }
+    }
+
+    fn try_from_nt(lwcid: nt::LightWeightConnectionId, bytes: &Bytes) -> Option<Self> {
+        let mut cbor = Deserializer::from(Cursor::new(bytes.deref()));
+        let msg_type: MessageType = cbor.deserialize().unwrap();
+        match msg_type {
+            MessageType::MsgGetHeaders => Some(ChainMessage::GetBlockHeaders(
+                lwcid,
+                cbor.deserialize_complete().unwrap(),
+            )),
+            MessageType::MsgHeaders => Some(ChainMessage::BlockHeaders(
+                lwcid,
+                cbor.deserialize_complete().unwrap(),
+            )),
+            MessageType::MsgGetBlocks => Some(ChainMessage::GetBlocks(
+                lwcid,
+                cbor.deserialize_complete().unwrap(),
+            )),
+            MessageType::MsgBlock => Some(ChainMessage::Block(
+                lwcid,
+                cbor.deserialize_complete().unwrap(),
+            )),
+            MessageType::MsgSubscribe1 => {
+                let v: u64 = cbor.deserialize_complete().unwrap();
+                let keep_alive = v == 43;
+                Some(ChainMessage::Subscribe(lwcid, keep_alive))
+            }
+            _ => unimplemented!(),
```

Comment:
This silently creates a crash on any accidentally unimplemented cases.
I'd rather add `unimplemented!()` cases on every variant in an exhaustive match.